src/rabbit_delayed_message.erl

```diff
@@ -154,84 +140,50 @@ code_change(_, State, _) -> {ok, State}.
-            not_set
-    end.
-
-route(#delay_key{exchange = Ex}, Deliveries, State) ->
+route_messages([], State) ->
+    State;
+route_messages([Key|Keys], #state{khepri_mod = Mod} = State) ->
+    {ok, Exchange} = Mod:get(Key++[exchange]),
+    [_, MsgKey] = Key,
+    V = rabbit_delayed_message_kv_store:do_take(MsgKey),
+    rabbit_delayed_message_kv_store:delete(MsgKey),
+    route(Exchange, [V], State),
+    Mod:delete(Key),
+    route_messages(Keys, State).
+
+route(Ex, Deliveries, State) ->
     ExName = Ex#exchange.name,
-    lists:map(fun (#delay_entry{delivery = Msg0}) ->
+    lists:map(fun (Msg0) ->
                       Msg1 = case Msg0 of
-                               #delivery{message = BasicMessage} ->
+                                 #delivery{message = BasicMessage} ->
                                      BasicMessage;
-                               _MC ->
-                                   Msg0
-                           end,
-                      Msg2 = swap_delay_header(Msg1),
+                                 _MC ->
+                                     Msg0
+                             end,
+                      Msg2 = rabbit_delayed_message_utils:swap_delay_header(Msg1),
                       Dests = rabbit_exchange:route(Ex, Msg2),
                       Qs = rabbit_amqqueue:lookup_many(Dests),
                       _ = rabbit_queue_type:deliver(Qs, Msg2, #{}, stateless),
                       bump_routed_stats(ExName, Qs, State)
               end, Deliveries).
 
-internal_delay_message(CurrTimer, Exchange, Message, Delay) ->
+internal_delay_message(#state{khepri_mod = Mod}, Exchange, Message, Delay) ->
     Now = erlang:system_time(milli_seconds),
-    %% keys are timestamps in milliseconds,in the future
     DelayTS = Now + Delay,
-    mnesia:dirty_write(?INDEX_TABLE_NAME,
-                       make_index(DelayTS, Exchange)),
-    mnesia:dirty_write(?TABLE_NAME,
-                       make_delay(DelayTS, Exchange, Message)),
-    case CurrTimer of
-        not_set ->
-            %% No timer in progress, so we start our own.
-            {ok, maybe_delay_first()};
-        _ ->
-            case erlang:read_timer(CurrTimer) of
-                false ->
-                    %% Timer is already expired.  Handler will be invoked soon.
-                    {ok, CurrTimer};
-                CurrMS when Delay < CurrMS ->
-                    %% Current timer lasts longer that new message delay
-                    _ = erlang:cancel_timer(CurrTimer),
-                    {ok, start_timer(Delay, make_key(DelayTS, Exchange))};
-                _ ->
-                    %% Timer is set to expire sooner than this
-                    %% message's scheduled delivery time.
-                    {ok, CurrTimer}
-            end
-    end.
+    Key = make_key(DelayTS, Exchange),
+    Mod:put([delayed_message_exchange, Key, delivery_time], DelayTS),
+    Mod:put([delayed_message_exchange, Key, exchange], Exchange),
```

Comment:
not sure how leveled works (I assume it does not keep all data in memory) although exchange is a metadata, to save some memory it could be stored in the kv_store along with the Message.
(Alternatively if exchange is stored in khepri, expired messages could be fetched grouped by exchange. Then looking up routing for an exchange can be done only once, sparing a few lookups (I think that's what the old implementation also did - might be just premature optimisation))